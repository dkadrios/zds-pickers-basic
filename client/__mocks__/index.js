export inputControls from './inputControls'
export mappings from './mappings'
export midi from './midi'
export shiftGroups from './shiftGroups'
export shifter from './shifter'
export store from './store'
export user from './user'
export version from './version'
