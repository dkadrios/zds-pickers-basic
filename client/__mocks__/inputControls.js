export default [
  {
    controlId: 0,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 11,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 1,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 1,
    status: 11,
    value: 111,
    active: false,
    lit: false,
  },
  {
    controlId: 2,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
  {
    controlId: 3,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 4,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: true,
  },
  {
    controlId: 5,
    channel: 10,
    calibrationLow: 0,
    calibrationHigh: 127,
    curve: 0,
    polarity: 0,
    latching: 0,
    status: 9,
    value: 110,
    active: true,
    lit: false,
  },
]
