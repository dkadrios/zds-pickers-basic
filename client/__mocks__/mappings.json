{
  "dialogVisible": false,
  "userDialogVisible": false,
  "userImportDialogVisible": false,
  "channels": [
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    null,
    "BFD3",
    null,
    null,
    null,
    null,
    null,
    null
  ],
  "stockMappings": [
    "2box",
    "Addictive Drums",
    "Alesis DM Pro",
    "Alesis DM10",
    "Alesis DM5",
    "BFD2",
    "BFD3",
    "Battery 3",
    "E-MU Protean Acoustic",
    "E-MU Protean Bread n Butter",
    "E-MU Protean Studio",
    "E-MU Protean Synth",
    "EZ Drummer The Classic EZX",
    "EZDrummer",
    "EZDrummer Jazz Kit",
    "EZDrummer Latin Percussion",
    "General MIDI",
    "Infinite Instruments",
    "John Emrich - Articulations 1",
    "Kontakt 2",
    "Ocean Way Drums",
    "Pearl Redbox",
    "Roland HandSonic 15",
    "Roland TD-10",
    "Roland TD-10 Perc 1",
    "Roland TD-10 Perc 2",
    "Roland TD-10 Perc 3",
    "Roland TD-10 Perc 4",
    "Roland TD-12",
    "Roland TD-12 Perc 1 (Latin)",
    "Roland TD-12 Perc 2 (Indian)",
    "Roland TD-12 Perc 3 (African)",
    "Roland TD-12 Perc 4 (Salsa)",
    "Roland TD-12 Perc 5 (SFX)",
    "Roland TD-12 Perc 6-8 (User)",
    "Roland TD-20",
    "Roland TD-20 Perc",
    "Roland TD-3",
    "Roland TD-30",
    "Roland TD-30 Perc 1 (Latin)",
    "Roland TD-30 Perc 2 (Indian)",
    "Roland TD-30 Perc 3 (African)",
    "Roland TD-30 Perc 4 (Salsa)",
    "Roland TD-30 Perc 5 (SFX)",
    "Roland TD-30 Perc 6-8 (User)",
    "Roland TD-5",
    "Roland TD-6",
    "Roland TD-6 Perc 1",
    "Roland TD-6 Perc 10 (Special)",
    "Roland TD-6 Perc 2",
    "Roland TD-6 Perc 3 (Room)",
    "Roland TD-6 Perc 4 (Power)",
    "Roland TD-6 Perc 5 (Elec)",
    "Roland TD-6 Perc 6 (808)",
    "Roland TD-6 Perc 7 (Jazz)",
    "Roland TD-6 Perc 8 (Brush)",
    "Roland TD-6 Perc 9 (Perc)",
    "Roland TD-7",
    "Roland TD-8",
    "Roland TD-8 Perc 1",
    "Roland TD-8 Perc 10 (Special)",
    "Roland TD-8 Perc 2",
    "Roland TD-8 Perc 3 (Room)",
    "Roland TD-8 Perc 4 (Power)",
    "Roland TD-8 Perc 5 (Elec)",
    "Roland TD-8 Perc 6 (808)",
    "Roland TD-8 Perc 7 (Jazz)",
    "Roland TD-8 Perc 8 (Brush)",
    "Roland TD-8 Perc 9 (Perc)",
    "Roland TD-9",
    "Roland TD-9 Perc",
    "Simmons SD1000",
    "Simmons SD5X",
    "Simmons SD7",
    "Simmons SD7PK",
    "Simmons SD9K",
    "Sonic Reality iMap Extended",
    "Sonic Reality iMap Standard",
    "Studio ProFiles",
    "Superior Drummer",
    "Yamaha DTX 700",
    "Yamaha DTXplorer",
    "Zendrum Stompblock",
    "Zendrum TRS",
    "Zendrum TRS 2.0"
  ],
  "userMappings": ["BFD3 Custom ZenEdit", "reaper"]
}
