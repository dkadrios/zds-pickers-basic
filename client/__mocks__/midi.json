{
  "devices": [
    {
      "id": "-1325431060",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "output",
      "version": "",
      "state": "connected",
      "connection": "open"
    },
    {
      "id": "-312145282",
      "manufacturer": "Zendrum Studio",
      "name": "ZDS Shifter",
      "type": "input",
      "version": "",
      "state": "connected",
      "connection": "open"
    }
  ],
  "listeningDevices": ["-312145282"]
}
