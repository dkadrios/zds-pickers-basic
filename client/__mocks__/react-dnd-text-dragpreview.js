/* eslint-disable import/prefer-default-export, jsx-a11y/alt-text */
import React from 'react'

export const createDragPreview = () => <img src={null} />
