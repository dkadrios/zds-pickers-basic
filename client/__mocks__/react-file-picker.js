/* eslint-disable import/prefer-default-export */
export const FilePicker = () => <div>content</div>
