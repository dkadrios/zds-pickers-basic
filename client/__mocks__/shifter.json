{
  "accessGranted": true,
  "errorMessage": "",
  "errorVisible": false,
  "exportBuffer": [],
  "exportDialogVisible": false,
  "exportFilename": "zds-shifter-backup.txt",
  "found": true,
  "hardwareTestVisible": true,
  "importDialogVisible": false,
  "importInProcess": false,
  "invalidSettingsFile": null,
  "midiActivityLEDMode": 0,
  "midiInActivity": false,
  "midiOutActivity": false,
  "ready": true,
  "resetInProcess": false,
  "responding": true,
  "searchedForShifter": true,
  "serialMidiOutEnabled": true,
  "showResetDialog": false,
  "testInterfaceFound": false,
  "usbMidiOutEnabled": true
}
