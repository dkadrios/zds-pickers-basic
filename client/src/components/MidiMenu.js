/* eslint-disable no-bitwise */
import { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import IconButton from '@material-ui/core/IconButton'
import ArrowDropDown from '@material-ui/icons/ArrowDropDown'
// import USB from '@material-ui/icons/Usb'
// import Plus from '@material-ui/icons/Add'
import Tooltip from '@material-ui/core/Tooltip'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
// import Midi from '../images/Midi.svg'
import { actions as shifterActions } from '../reducers/shifter'
import { shifterShape } from '../core/shapes'

const options = [
  { value: 4, label: 'Enable MIDI Out on DIN', tag: 'miDinOut' },
  { value: 8, label: 'Enable MIDI Out on USB', tag: 'miUsbOut' },
  { value: 12, label: 'Enable Both DIN AND USB', tag: 'miBothOut' },
]

export class MidiMenu extends Component {
  // eslint-disable-next-line react/state-in-constructor
  state = {
    anchorEl: null,
  }

  handleClick = (event) => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleMenuItemClick = (value) => {
    const { setFlags, shifter } = this.props
    const { midiActivityLEDMode } = shifter
    const serialMidiOutEnabled = Boolean(value & 4)
    const usbMidiOutEnabled = Boolean(value & 8)
    setFlags(midiActivityLEDMode, serialMidiOutEnabled, usbMidiOutEnabled)
    this.handleClose()
  }

  handleClose = () => {
    this.setState({ anchorEl: null })
  }

  render() {
    const { anchorEl } = this.state
    const { shifter: { ready, found, responding, serialMidiOutEnabled, usbMidiOutEnabled } } = this.props

    const isValueSelected = (value) => {
      let check = 0
      if (serialMidiOutEnabled) {
        check |= 4
      }
      if (usbMidiOutEnabled) {
        check |= 8
      }
      return value === check
    }

    const menuLabel = () => serialMidiOutEnabled && usbMidiOutEnabled ? 'Both DIN and USB' : serialMidiOutEnabled ? 'DIN Only' : 'USB Only'

    return ready ? (
      <>
        <IconButton
          aria-haspopup="true"
          aria-owns={anchorEl ? 'simple-menu' : null}
          disabled={!(found && responding)}
          onClick={this.handleClick}
        >
          <Tooltip
            enterDelay={250}
            placement="top"
            title="MIDI out mode"
          >
            <div style={{ display: 'flex', flexFlow: 'row nowrap', alignItems: 'center' }}>
              {/* {serialMidiOutEnabled && <Midi width={18} />}
              {serialMidiOutEnabled && usbMidiOutEnabled && (
                <Plus style={{ width: 15, position: 'relative', left: 3 }} />
              )}
              {usbMidiOutEnabled && <USB />} */}
              TODO:
              <span style={{ padding: '5px 5px 0' }}>{menuLabel()}</span>
              <ArrowDropDown />
            </div>
          </Tooltip>
        </IconButton>

        <Menu
          anchorEl={anchorEl}
          id="simple-menu"
          onClose={this.handleClose}
          open={Boolean(anchorEl)}
        >
          {options.map(({ value, label, ...rest }) => (
            <MenuItem
              key={value}
              onClick={() => this.handleMenuItemClick(value)}
              selected={isValueSelected(value)}
              {...rest}
            >
              {label}
            </MenuItem>
          ))}
        </Menu>
      </>
    ) : null
  }
}

MidiMenu.propTypes = {
  shifter: shifterShape.isRequired,
  setFlags: PropTypes.func.isRequired,
}

export const mapStateToProps = ({ shifter }) => ({ shifter })
export const mapDispatchToProps = dispatch => bindActionCreators(shifterActions, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MidiMenu)
