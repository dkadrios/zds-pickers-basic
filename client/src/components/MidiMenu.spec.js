import flow from 'lodash/fp/flow'
import set from 'lodash/fp/set'
import { shifter } from '../../__mocks__'
import { MidiMenu, mapDispatchToProps, mapStateToProps } from './MidiMenu'

describe('MidiMenu tests', () => {
  const setFlags = jest.fn()
  const props = {
    shifter,
    setFlags,
  }

  beforeEach(jest.clearAllMocks)

  it('renders correctly', () => {
    shallowExpect(MidiMenu)(props).toMatchSnapshot()
  })

  it('renders correctly in alternate modes', () => {
    shallowExpect(MidiMenu)(flow(
      set('shifter.serialMidiOutEnabled', false),
      set('shifter.usbMidiOutEnabled', false),
    )(props)).toMatchSnapshot()
  })

  it('clicking should open the menu', () => {
    const wrapper = shallow(<MidiMenu {...props} />)
    wrapper.find('WithStyles(IconButton)').simulate('click', { currentTarget: 'target' })
    expect(wrapper.state('anchorEl')).toEqual('target')
  })

  it('clicking should hide the menu when closing', () => {
    const wrapper = shallow(<MidiMenu {...props} />)
    wrapper.setState({ anchorEl: 'target' })
    wrapper.instance().handleClose()
    expect(wrapper.state('anchorEl')).toBeNull()
  })

  it('should update flags when items miDinOut is clicked', () => {
    const wrapper = shallow(<MidiMenu {...props} />)
    wrapper
      .find('WithStyles(MenuItem)')
      .filter('[tag="miDinOut"]')
      .simulate('click')
    expect(setFlags).toHaveBeenCalledWith(shifter.midiActivityLEDMode, true, false)
  })

  it('should update flags when items miUsbOut is clicked', () => {
    const wrapper = shallow(<MidiMenu {...props} />)
    wrapper
      .find('WithStyles(MenuItem)')
      .filter('[tag="miUsbOut"]')
      .simulate('click')
    expect(setFlags).toHaveBeenCalledWith(shifter.midiActivityLEDMode, false, true)
  })

  it('should update flags when items miBothOut is clicked', () => {
    const wrapper = shallow(<MidiMenu {...props} />)
    wrapper
      .find('WithStyles(MenuItem)')
      .filter('[tag="miBothOut"]')
      .simulate('click')
    expect(setFlags).toHaveBeenCalledWith(shifter.midiActivityLEDMode, true, true)
  })

  it('should utilize mapStateToProps', () => {
    const state = { shifter }
    expect(mapStateToProps(state)).toEqual(state)
  })

  it('should utilize mapDispatchToProps', () => {
    const dispatch = jest.fn()
    mapDispatchToProps(dispatch).setFlags(3, true, false)
    expect(dispatch.mock.calls[0][0]).toEqual({
      midiActivityLEDMode: 3,
      serialMidiOutEnabled: true,
      type: 'SET_FLAGS',
      usbMidiOutEnabled: false,
    })
  })
})
