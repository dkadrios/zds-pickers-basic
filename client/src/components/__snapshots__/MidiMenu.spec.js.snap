// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MidiMenu tests renders correctly 1`] = `
<Fragment>
  <WithStyles(IconButton)
    aria-haspopup="true"
    aria-owns={null}
    disabled={false}
    onClick={[Function]}
  >
    <WithStyles(Tooltip)
      enterDelay={250}
      placement="top"
      title="MIDI out mode"
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexFlow": "row nowrap",
          }
        }
      >
        <test-file-stub
          width={18}
        />
        <pure(AddIcon)
          style={
            Object {
              "left": 3,
              "position": "relative",
              "width": 15,
            }
          }
        />
        <pure(UsbIcon) />
        <span
          style={
            Object {
              "padding": "5px 5px 0",
            }
          }
        >
          Both DIN and USB
        </span>
        <pure(ArrowDropDownIcon) />
      </div>
    </WithStyles(Tooltip)>
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl={null}
    id="simple-menu"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(MenuItem)
      key="4"
      onClick={[Function]}
      selected={false}
      tag="miDinOut"
    >
      Enable MIDI Out on DIN
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="8"
      onClick={[Function]}
      selected={false}
      tag="miUsbOut"
    >
      Enable MIDI Out on USB
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="12"
      onClick={[Function]}
      selected={true}
      tag="miBothOut"
    >
      Enable Both DIN AND USB
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
</Fragment>
`;

exports[`MidiMenu tests renders correctly in alternate modes 1`] = `
<Fragment>
  <WithStyles(IconButton)
    aria-haspopup="true"
    aria-owns={null}
    disabled={false}
    onClick={[Function]}
  >
    <WithStyles(Tooltip)
      enterDelay={250}
      placement="top"
      title="MIDI out mode"
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexFlow": "row nowrap",
          }
        }
      >
        <span
          style={
            Object {
              "padding": "5px 5px 0",
            }
          }
        >
          USB Only
        </span>
        <pure(ArrowDropDownIcon) />
      </div>
    </WithStyles(Tooltip)>
  </WithStyles(IconButton)>
  <WithStyles(Menu)
    anchorEl={null}
    id="simple-menu"
    onClose={[Function]}
    open={false}
  >
    <WithStyles(MenuItem)
      key="4"
      onClick={[Function]}
      selected={false}
      tag="miDinOut"
    >
      Enable MIDI Out on DIN
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="8"
      onClick={[Function]}
      selected={false}
      tag="miUsbOut"
    >
      Enable MIDI Out on USB
    </WithStyles(MenuItem)>
    <WithStyles(MenuItem)
      key="12"
      onClick={[Function]}
      selected={false}
      tag="miBothOut"
    >
      Enable Both DIN AND USB
    </WithStyles(MenuItem)>
  </WithStyles(Menu)>
</Fragment>
`;
